### PR TITLE
Move save_hw_clock method to a D-Bus configuration task

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

**Comment by KKoukiou** [comment]
I have troubles writing a 'successful' scenario test. I think there was none before. Do you think it's that possible to have one?

File: pyanaconda/installation.py [hunk 1]
```diff
@@ -253,13 +252,6 @@ def _prepare_installation(payload, ksdata):
     installation_queue.queue_started.connect(lambda x: progress_message(x.status_message))
     installation_queue.task_completed.connect(lambda x: progress_step(x.name))
 
-    # Save system time to HW clock.
-    # - this used to be before waiting on threads, but I don't think that's needed
-    if conf.system.can_set_hardware_clock:
```
Comment (poncovka) [inline pyanaconda/installation.py:258]: Please, check this condition in the `run` method of the new task.

File: pyanaconda/modules/timezone/installation.py [hunk 2]
```diff
@@ -187,3 +187,37 @@ def _write_configuration(self):
...
+
+    @property
+    def name(self):
+        return "Set the Hardware Clock from the System Clock"
+
+    def run(self):
+        """Perform the actual work of setting the Hardware Clock from the System Clock."""
+        self._save_hw_clock()
+
+    def _save_hw_clock(self):
```
Comment (poncovka) [inline pyanaconda/modules/timezone/installation.py:211]: I would merge the `_save_hw_clock` method with the `run` method.

File: pyanaconda/modules/timezone/installation.py [hunk 3]
```diff
@@ -187,3 +187,37 @@ def _write_configuration(self):
...
+        if arch.is_s390():
+            # there is no HW clock on s390(x)
```
Comment (poncovka) [inline pyanaconda/modules/timezone/installation.py:213]: Could you log a message about skipping this task on s390x, please?

File: pyanaconda/modules/timezone/installation.py [hunk 4]
```diff
@@ -187,3 +187,37 @@ def _write_configuration(self):
             except ntp.NTPconfigError as ntperr:
                 log.warning("Failed to save NTP configuration without chrony package: %s",
                             ntperr)
+
+
+class ConfigureHwClockTask(Task):
```
Comment (poncovka) [inline pyanaconda/modules/timezone/installation.py:192]: I would prefer `ConfigureHardwareClockTask` for better clarity.

File: pyanaconda/modules/timezone/timezone.py [hunk 5]
```diff
@@ -210,5 +211,8 @@ def install_with_tasks(self):
                 sysroot=conf.target.system_root,
                 ntp_enabled=self.ntp_enabled,
                 ntp_servers=self.time_sources
+            ),
+            ConfigureHwClockTask(
```
Comment (poncovka) [inline pyanaconda/modules/timezone/timezone.py:215]: Could you move the task on the top of the list, please? We are already changing its position, so let's keep at least the order of these tasks.

File: tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py [hunk 6]
```diff
@@ -587,3 +598,22 @@ def _validate_ntp_config(self, sysroot, was_present, was_enabled, expected_lines
...
+    """Test the D-Bus Timezone Hardware Clock task."""
+
+    @patch("pyanaconda.modules.timezone.installation.util.execWithRedirect")
+    @patch('pyanaconda.modules.timezone.installation.arch.is_s390', return_value=True)
+    def test_s390_save_hw_clock(self, mock_is_s390, mock_exec_with_redirect):
+        """Test that save_hw_clock does nothing on s390."""
+        self._execute_task(False)
+        # expected state: calling it only once in the check for architecture
+        mock_is_s390.assert_called_once()
+        mock_exec_with_redirect.assert_not_called()
```
Comment (poncovka) [inline tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py:613]: Could you please add tests that will check the `hwclock --utc` call and `hwclock --local` call? It should be mostly a copy-paste of this test.

**Review by poncovka — COMMENTED** [review]
I have some smaller suggestions, otherwise, it looks great!

**Comment by poncovka** [comment]
> I have troubles writing a 'successful' scenario test. I think there was none before. Do you think it's that possible to have one?

It should be straightforward, but I will be happy to help you with that :-)

**Review by VladimirSlavik — COMMENTED** [review]
Not much to add after Vendy's review. It's going in the right direction.

File: tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py [hunk 7]
```diff
@@ -290,19 +295,24 @@ def test_install_with_tasks_configured(self, publisher):
...
         task_paths = self.timezone_interface.InstallWithTasks()
         task_objs = check_task_creation_list(task_paths, publisher, task_classes)
 
+        # ConfigureHardwareClockTask
+        obj = task_objs[1]
+        assert obj.implementation._is_utc is True
+
         # ConfigureTimezoneTask
-        obj = task_objs[0]
+        obj = task_objs[1]
```
Comment (VladimirSlavik) [inline tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py:310]: You have twice `[1]` but no `[0]` here :)

**Review by VladimirSlavik — APPROVED** [review]
Vendy can see more than I do, so I may not be the last word here, but I think it's perfect now :)

File: tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py (hunk 7 above)
Comment (KKoukiou) [reply tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py]: Oops, thanks!

File: tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py [hunk 8]
```diff
@@ -587,3 +597,44 @@ def _validate_ntp_config(self, sysroot, was_present, was_enabled, expected_lines
...
+        self._execute_task(False)
+        # expected state: calling it only once in the check for architecture
+        mock_is_s390.assert_called_once()
+        mock_exec_with_redirect.assert_not_called()
+
+    @patch("pyanaconda.modules.timezone.installation.util.execWithRedirect")
+    def test_hwclock_config_task_local(self, mock_exec_with_redirect):
+        """Test the Hardware clock configuration task - local"""
+        with patch("pyanaconda.modules.timezone.installation.conf") as core_util:
+            core_util.system.can_set_hardware_clock(True)
```
Comment (poncovka) [inline tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py:618]: I think that `conf.system.can_set_hardware_clock` still returns the default mock object after this call. Try: `core_util.system.can_set_hardware_clock = True`

Comment (poncovka) [reply tests/unit_tests/pyanaconda_tests/modules/timezone/test_module_timezone.py:618]: Btw. you should be able to add a test for `conf.system.can_set_hardware_clock == False` now.

**Review by poncovka — APPROVED** [review]
There are some small issues in the tests, otherwise, it looks great :-)

**Review by poncovka — APPROVED** [review]
Thanks!

**Pull request merged**